### PR TITLE
OSDOCS#10250: Added information around additional principals

### PR DESCRIPTION
**Pull request by EricPonvelle**
Version(s):
`enterprise-4.16`+

Issue:
**[OSDOCS-10250](https://issues.redhat.com/browse/OSDOCS-10250)**

Link to docs preview:
- **[`create` object](https://file.rdu.redhat.com/eponvell/OSDOCS-10250_Additional-Principals/cli_reference/rosa_cli/rosa-manage-objects-cli.html#rosa-create-cluster-command_rosa-managing-objects-cli)** updates
- **[`edit` object](https://file.rdu.redhat.com/eponvell/OSDOCS-10250_Additional-Principals/cli_reference/rosa_cli/rosa-manage-objects-cli.html#rosa-edit-cluster_rosa-managing-objects-cli)** updates
- Adding additional principals to your cluster (new)
  - **[Overview](https://77936--ocpdocs-pr.netlify.app/openshift-rosa/latest/rosa_hcp/rosa-hcp-aws-private-creating-cluster.html#rosa-additional-principals-overview_rosa-hcp-aws-private-creating-cluster)**
    image
  - **[On a new cluster](https://77936--ocpdocs-pr.netlify.app/openshift-rosa/latest/rosa_hcp/rosa-hcp-aws-private-creating-cluster.html#rosa-additional-principals-create_rosa-hcp-aws-private-creating-cluster)**
  - **[On an existing cluster](https://77936--ocpdocs-pr.netlify.app/openshift-rosa/latest/rosa_hcp/rosa-hcp-aws-private-creating-cluster.html#rosa-additional-principals-edit_rosa-hcp-aws-private-creating-cluster)**

QE review:
- [x] QE has approved this change.

Additional information:
This PR adds arguments and instructions about adding additional principals to your cluster's trust policy.

### CODE REVIEW

**Comment by bchandra-ocp** [comment]
Hi Eric, the whole feature documentation is under "Creating ROSA with HCP clusters using the default options". Adding allowed principal is not among default option - it's optional. so, how do we address? Here are couple of suggestions:
1. Create a new topic for Creating .... with optional configurations
2. Modify the existing title to 'Creating ROSA with HCP clusters' and have two subsections respectively for Default configurations and custom configurations. [I don't prefer calling 'default options' because the phrase is too generic']

File: modules/rosa-additional-principals-overview.adoc [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+
+
+:_mod-docs-content-type: CONCEPT
+[id="rosa-additional-principals-overview_{context}"]
+= Additional principals on your ROSA cluster
+
+You can approve additional principals to connect to your cluster. You can add the following principal types:
```
Comment (bchandra-ocp) [inline modules/rosa-additional-principals-overview.adoc:11]: "You can allow AWS IAM Roles as additional principals to connect to your cluster's private API Server endpoint."

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]

**Comment by bchandra-ocp** [comment]
We need a couple of sentences to provide an overview of the feature covering what and why before jumping into the details. So may I suggest following be added at the top/beginning of the documentation for this feature? 

**ROSA with HCP cluster's API Server endpoint can be accessed either via the public internet or via the interface endpoint created within the VPC private subnets where clusters worker nodes are deployed. By default, the HCP API Server is privately accessed using the operator role <prefix>-kube-system-kube-controller-manager. In order to access the HCP API Server from another trusted AWS account directly and privately, you must include IAM Roles as additional allowed principals.  Using this feature, you can avoid peering or attaching additional VPCs to cluster's VPC thereby simplifying network architecture and reducing data transfer costs.**

File: modules/rosa-additional-principals-create.adoc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+:_mod-docs-content-type: PROCEDURE
+[id="rosa-additional-principals-create_{context}"]
+= Adding additional principals while creating your ROSA cluster
+
+Use the `--additional-allowed-principals` argument to permit access through other accounts and roles.
+
+.Procedure
+
+. In your terminal, add `--additional-allowed-principals` to the `rosa create cluster` command, similar to the following:
```
Comment (bchandra-ocp) [inline modules/rosa-additional-principals-create.adoc:15]: suggest rephrasing as:

Use --additional-allowed-principals field in the 'rosa create cluster' CLI command, as shown below:

File: modules/rosa-additional-principals-create.adoc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
++
+[source,terminal]
+----
+$ rosa create cluster [...] --additional-allowed-principals <arn_string>
+----
++
+You can use `arn:aws:iam::account_id:root` to approve an account, `arn:aws:iam::account_id:role/role_name` to a approve a specific role, `arn:aws:iam::account_id:user/user_name` to approve an additional user, and `arn:aws:iam::account_id:federated-user/user_name` to approve a federated user.
```
Comment (bchandra-ocp) [inline modules/rosa-additional-principals-create.adoc:22]: Again - I think we only support /role so kindly check.

Comment (EricPonvelle) [reply modules/rosa-additional-principals-create.adoc:22]: When I tested it, it attached with a user role, though I wasn't sure how to test beyond that. Do you know who would be our QE contact for this? Maybe I can get access to their test plan.

**Comment by EricPonvelle** [comment]
@bchandra-ocp I updated the PR with your suggestions. I'll create a new topic for creating ROSA with HCP topics with customizations, and this can be the first section there. Could we also add this to the authentication book?

File: modules/rosa-additional-principals-create.adoc (hunk 3 above)
Comment (EricPonvelle) [reply modules/rosa-additional-principals-create.adoc:22]: @bchandra-ocp confirmed it is just user roles. I've updated the docs.

`W: The current version (1.2.41-RC1) is not up to date with latest released version (1.2.41).`
`W: It is recommended that you update to the latest version.`
`E: Failed to update cluster: Principal 'arn:aws:iam::123456789012:root' must be of format:`
`'arn:aws:iam::123456789012:role/name'`

File: modules/rosa-additional-principals-overview.adoc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+:_mod-docs-content-type: CONCEPT
+[id="rosa-additional-principals-overview_{context}"]
+= Additional principals on your ROSA cluster
+
+You can allow AWS IAM Roles as additional principals to connect to your cluster's private API server endpoint.
+
+You can access your {hcp-title} cluster's API Server endpoint from either the public internet or the interface endpoint that was created within the VPC private subnets. By default, you can privately access your {hcp-title} API Server by using the `-kube-system-kube-controller-manager` Operator role. You must include IAM Roles from coss-account as additional allowed principals to be able to allow {hcp-title} API servce access by using another trusted AWS account. This feature offers a way to simplify your network architecture and reduced transfer costs as well as avoiding peering or attaching additional VPCs to cluster's VPC.
```
Comment (bchandra-ocp) [inline modules/rosa-additional-principals-overview.adoc:11]: Typo... _cross-account_

Comment (bchandra-ocp) [reply modules/rosa-additional-principals-overview.adoc:11]: Which one reads well
You must include IAM Roles from coss-account as additional allowed principals to be able to allow ROSA with HCP API servce access by using another trusted AWS account.

To be able to access ROSA with HCP API server from another account directly without using the primary account where cluster is installed, you must include cross-account IAM Roles as additional principals.

Comment (bchandra-ocp) [reply modules/rosa-additional-principals-overview.adoc:11]: Typo... HCP API _server_

Comment (bchandra-ocp) [reply modules/rosa-additional-principals-overview.adoc:11]: _reduced transfer costs as well as avoiding peering or attaching additional VPCs to cluster’s VPC._

This feature allows you to simplify your network architecture and reduce data transfer costs by avoiding peering or attaching cross-account VPCs to cluster's VPC.

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]

File: modules/rosa-additional-principals-overview.adoc [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+:_mod-docs-content-type: CONCEPT
+[id="rosa-additional-principals-overview_{context}"]
+= Additional principals on your ROSA cluster
+
+You can allow AWS IAM Roles as additional principals to connect to your cluster's private API server endpoint.
+
+You can access your {hcp-title} cluster's API Server endpoint from either the public internet or the interface endpoint that was created within the VPC private subnets. By default, you can privately access your {hcp-title} API Server by using the `-kube-system-kube-controller-manager` Operator role. To be able to access ROSA with HCP API server from another account directly without using the primary account where cluster is installed, you must include cross-account IAM Roles as additional principals. This feature allows you to simplify your network architecture and reduce data transfer costs by avoiding peering or attaching cross-account VPCs to cluster's VPC.
```
Comment (jneczypor) [inline modules/rosa-additional-principals-overview.adoc:11]:
```suggestion
You can access your {hcp-title} cluster's API Server endpoint from either the public internet or the interface endpoint that was created within the VPC private subnets. By default, you can privately access your {hcp-title} API Server by using the `-kube-system-kube-controller-manager` Operator role. To be able to access ROSA with HCP API server from another account directly without using the primary account where cluster is installed, you must include cross-account IAM roles as additional principals. This feature allows you to simplify your network architecture and reduce data transfer costs by avoiding peering or attaching cross-account VPCs to cluster's VPC.
```

File: modules/rosa-create-objects.adoc [hunk 6]
```diff
@@ -192,6 +192,9 @@ $ rosa create cluster --cluster-name=<cluster_name> [arguments]
 |--additional-control-plane-security-group-ids <sec_group_id>
 |The identifier of one or more additional security groups to use along with the default security groups that are used with the control plane nodes created alongside the cluster. For more information on additional security groups, see the requirements for _Security groups_ under _Additional resources_.
 
+|--additional-allowed-principals <arn>
+|A comma-separated list of additional allowed principal ARNs to be added to the Hosted Control Plane's VPC Endpoint Service to enable additional VPC Endpoint connection requests to be automatically accepted.
```
Comment (jneczypor) [inline modules/rosa-create-objects.adoc:196]:
```suggestion
|A comma-separated list of additional allowed principal ARNs to be added to the hosted control plane's VPC Endpoint Service to enable additional VPC Endpoint connection requests to be automatically accepted.
```
Check to make sure that "endpoint service" and "endpoint" need to be capitalized.

File: modules/rosa-edit-objects.adoc [hunk 7]
```diff
@@ -25,6 +25,9 @@ $ rosa edit cluster --cluster=<cluster_name> | <cluster_id> [arguments]
 |===
 |Option |Definition
 
+|--additional-allowed-principals <arn>
+|A comma-separated list of additional allowed principal ARNs to be added to the Hosted Control Plane's VPC Endpoint Service to enable additional VPC Endpoint connection requests to be automatically accepted.
```
Comment (jneczypor) [inline modules/rosa-edit-objects.adoc:29]: See same notes as above.

**Review by jneczypor — COMMENTED** [review]
Just a few changes.

**Comment by bchandra-ocp** [comment]
For the sake for completeness, we'd have to include a "Note" for the feature calling out steps customer has to take in their AWS cloud account after adding additional principals. Let me propose text .

File: modules/rosa-additional-principals-overview.adoc [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+image::AWS_cross_account_access.png[Overview of AWS cross account access]
+
+In this diagram, the cluster creating account is designated as Account A. This account designates that another account, Account B, should have access to the API server.
```
Comment (bchandra-ocp) [inline modules/rosa-additional-principals-overview.adoc:15]: _suggesting to include following note_

Once configuring additional allowed principals, you will need to create Interface VPC Endpoint in the VPC from where you want to access the cross-account ROSA with HCP API Server. Then, create private hosted zone in Route53 to route calls made to cross-account ROSA with HCP API Server to pass through the created VPC Endpoint.

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]

File: modules/rosa-additional-principals-create.adoc [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+. Add the `--additional-allowed-principals` argument to the `rosa create cluster` command, similar to the following:
++
+[source,terminal]
+----
+$ rosa create cluster [...] --additional-allowed-principals <arn_string>
+----
++
+You can use `arn:aws:iam::account_id:role/role_name` to a approve a specific role.
+
+. When the cluster creation command executes, you receive a summary of your cluster with the `--additional-allowed-principals` specified:
```
Comment (lpettyjo) [inline modules/rosa-additional-principals-create.adoc:24]: According to IBM SG, you should use "run" rather than "execute", if possible.

**Review by lpettyjo — COMMENTED** [review]
Otherwise, LGTM!

File: modules/rosa-additional-principals-create.adoc [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+.Procedure
+
+. Add the `--additional-allowed-principals` argument to the `rosa create cluster` command, similar to the following:
++
+[source,terminal]
+----
+$ rosa create cluster [...] --additional-allowed-principals <arn_string>
+----
++
+You can use `arn:aws:iam::account_id:role/role_name` to a approve a specific role.
```
Comment (aaraj7) [inline modules/rosa-additional-principals-create.adoc:22]:
```suggestion
You can use `arn:aws:iam::account_id:role/role_name` to approve a specific role.
```

File: modules/rosa-additional-principals-create.adoc [hunk 11]
```diff
@@ -0,0 +1,83 @@
+// Module included in the following assemblies:
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+
+
+:_mod-docs-content-type: PROCEDURE
+[id="rosa-additional-principals-create_{context}"]
+= Adding additional principals while creating your ROSA cluster
```
Comment (aaraj7) [inline modules/rosa-additional-principals-create.adoc:9]: What about changing the title to ROSA HCP cluster? as ROSA classic cluster doesn't support this feature

File: modules/rosa-additional-principals-edit.adoc [hunk 12]
```diff
@@ -0,0 +1,22 @@
+// Module included in the following assemblies:
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+
+
+:_mod-docs-content-type: PROCEDURE
+[id="rosa-additional-principals-edit_{context}"]
+= Adding additional principals to your existing ROSA cluster
```
Comment (aaraj7) [inline modules/rosa-additional-principals-edit.adoc:9]:
```suggestion
= Adding additional principals to your existing ROSA with HCP cluster
```

File: modules/rosa-additional-principals-overview.adoc [hunk 13]
```diff
@@ -0,0 +1,20 @@
+// Module included in the following assemblies:
+//
+// * rosa_install_access_delete_clusters/rosa_getting_started_iam/rosa-accessing-cluster.adoc
+// * rosa_install_access_delete_clusters/rosa-sts-accessing-cluster.adoc
+:_mod-docs-content-type: CONCEPT
+[id="rosa-additional-principals-overview_{context}"]
+= Additional principals on your ROSA cluster
```
Comment (aaraj7) [inline modules/rosa-additional-principals-overview.adoc:7]:
```suggestion
= Additional principals on your ROSA with HCP cluster
```

**Pull request merged**